src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_TypeInference.cs

```diff
@@ -0,0 +1,1341 @@
+                //         y = null;
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(14, 13));
+        }
+
+        // PROTOTYPE(NullableReferenceTypes): Deconstruction declaration ignores nullability.
+        [Fact(Skip = "TODO")]
+        public void DeconstructionTypeInference_04()
+        {
+            var source =
+@"using System;
+using System.Collections.Generic;
+class C
+{
+    static IEnumerable<(string, string?)> F() => throw new Exception();
+    static void G()
+    {
+        foreach ((var x, var y) in F())
+        {
+            x.ToString();
+            y.ToString();
+        }
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8602: Possible dereference of a null reference.
+                //             y.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(11, 13));
+        }
+
+        // PROTOTYPE(NullableReferenceTypes): Type inference incorrect.
+        [Fact(Skip = "TODO")]
+        public void TypeInference_TupleNameDifferences_01()
+        {
+            var source =
+@"class C<T>
+{
+}
+static class E
+{
+    public static T F<T>(this C<T> c, T t) => t;
+}
+class C
+{
+    static void F(object o)
+    {
+        var c = new C<(object x, int y)>();
+        c.F((o, -1)).x.ToString();
+    }
+}";
+
+            var comp = CreateCompilationWithMscorlibAndSystemCore(
+                source,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
+                parseOptions: TestOptions.Regular7);
+            comp.VerifyDiagnostics(
+                // (13,22): error CS1061: '(object, int)' does not contain a definition for 'x' and no extension method 'x' accepting a first argument of type '(object, int)' could be found (are you missing a using directive or an assembly reference?)
+                //         c.F((o, -1)).x.ToString();
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "x").WithArguments("(object, int)", "x").WithLocation(13, 22));
+
+            comp = CreateCompilationWithMscorlibAndSystemCore(
+                source,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (13,22): error CS1061: '(object, int)' does not contain a definition for 'x' and no extension method 'x' accepting a first argument of type '(object, int)' could be found (are you missing a using directive or an assembly reference?)
+                //         c.F((o, -1)).x.ToString();
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "x").WithArguments("(object, int)", "x").WithLocation(13, 22));
+        }
```

Comment:
May be worthwhile to cover similar combinations of nullability and dynamic or modopts.

Reply:
There are several tests for nullability and dynamic below.